Rewrite of the QUIC layer.

sbordet opened the pull request:
Rewrite of the QUIC layer to accommodate, in the future, implementations alternative to Quiche.

sbordet committed:
Rewrite of the QUIC layer for the implementation based on the Quiche library.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet commented:
The main work was to remove the hardcoded dependencies on Quiche.

For this reason, module `jetty-quic-api` is introduced with `Session` and `Stream` similar to HTTP/2 and HTTP/3.
Along with these classes, the QUIC frames defined by the spec, that appear in method signatures of `Session` and `Stream`.

The Quiche implementation is now an implementation of `jetty-quic-api`.

The HTTP/3 layer now only depends on `jetty-quic-api`.

sbordet committed:
Removed leftovers of quic-streams.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Merged branch 'jetty-12.1.x' into 'fix/12.1.x/quiche-rewrite'.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Fixed Jetty modules and DistributionTest.testH3().

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Fixed ServerTimeoutsTest.testIdleTimeout().

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Dependency on jetty-quic-quiche-foreign only for Java 22+.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Fixed Http3AsyncIOServletTest in ee9-10-11.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Merged branch 'jetty-12.1.x' into 'fix/12.1.x/quiche-rewrite'.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Fixed QuicheSession.disconnect() by waiting for the Flusher.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

gregw commented on jetty-core/jetty-quic/jetty-quic-api/src/main/java/org/eclipse/jetty/quic/api/Stream.java:
```diff
@@ -0,0 +1,427 @@
+     * @return the {@link CompletableFuture} that gets notified when the
+     * frame has been sent
+     */
+    CompletableFuture<Stream> data(boolean last, ByteBuffer... data);
```
Use `Callback` rather than CF

gregw commented on jetty-core/jetty-quic/jetty-quic-common/src/main/java/org/eclipse/jetty/quic/common/StreamEndPoint.java:
```diff
@@ -0,0 +1,630 @@
+        };
+    }
+
+    public void write(Callback callback, List<ByteBuffer> buffers, boolean last)
```
think about cancelling writes

sbordet committed:
Replaced APIs that used CompletableFuture with Promise.Invocable.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Promise cleanup and documentation.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Fixed StreamEndPoint.write() overrides.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Preserved InvocationType for BiConsumer implementations.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Merged branch 'jetty-12.1.x' into 'jetty-12.1.x-quic-streams'.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Fixed test expectations.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Fixed logging that was causing NPEs.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Made StreamEndPoint callback an Invocable task passed to the QUIC session AES.

Made Quiche.drainClearBytesForStream() return the FIN flag indication.

Improved Qpack[Decoder|Encoder] locking.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Merged branch 'jetty-12.1.x' into 'jetty-12.1.x-quic-streams'.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Fixed logic in case of BLOCKED_FRAME.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Merged branch 'jetty-12.1.x' into 'jetty-12.1.x-quic-streams'.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Fixed logic in case of BLOCKED_FRAME for trailers.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Merged branch 'jetty-12.1.x' into 'jetty-12.1.x-quic-streams'.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Reconnected InvocationType for the server side.

Made relevant H3 unidirectional connections non-blocking.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Reduced concurrency in HttpClientLoadTest.testConcurrent().

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Improved code comment.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Removed HTTP3StreamServer.dataReady, as it is not necessary.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Fixed Origin.isSecure() for "wss".

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Updated documentation.

Removed necessity of passing the QuicheClientQuicConfiguration to QuicheTransport, so it can now be a singleton.

Updated server module to depend on virtual http3-impl, with a default provider for http3-quiche.

Enabled ee11 client transport tests in CI.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Removed TODO.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Reverted removal of ClientQuicConfiguration from QuicheTransport, as it is necessary when QuicheTransport is used by non-HTTP/3 protocols.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Fixed re-enabled ee11 client transport tests for H3_QUICHE.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Removed the need for BooleanSupplier in MessageParser.
Now relying on the returned boolean from Quiche.drainClearBytesForStream().

Updated HTTP3StreamConnection to avoid copying the HTTP/3 bytes, and use the returned boolean to generate HTTP/3 frames.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Improved test case reporting.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Fixed AsyncIOServletTest.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Fixed release of QUIC Stream.Data.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Implemented support for minInputBufferSpace in QuicheStream.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Improved RetainableByteBuffer reuse in QuicheStream.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Fixed flaky test GoAwayTest.testServerGoAwayWithInFlightStreamClientFailsStream().

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Merged branch 'jetty-12.1.x' into 'jetty-12.1.x-quic-streams'.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Removed getOrCreate*() methods in favor of explicit get*() and create*().

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Merged branch 'jetty-12.1.x' into 'jetty-12.1.x-quic-streams'.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Introduced ParserListener as explicit different object to receive HTTP/3 parser events, rather than implicitly be HTTP3Session.
This allows for customization of the parser events processing, for example serializing them.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Reverted logic in StreamEndPoint where notification of readability of the QUIC Stream was parallelized: now it is serialized.
The reason was to avoid that for HTTP/3 we had to serialize just after having parallelized, due to the fact that control frames and message frames can influence the HTTP/3 session state.

Improved exception throwing and handling.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Merged branch 'jetty-12.1.x' into 'jetty-12.1.x-quic-streams'.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Merged branch 'jetty-12.1.x' into 'jetty-12.1.x-quic-streams'.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Implemented HttpStreamOverHTTP3.cancelSend().

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Updated FrameGenerator.generate() to return long instead of int.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Merged branch 'jetty-12.1.x' into 'jetty-12.1.x-quic-streams'.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Merged branch 'jetty-12.1.x' into 'jetty-12.1.x-quic-streams'.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

gregw commented:
@sbordet status?

sbordet commented:
@gregw to be reviewed

gregw commented:
@sbordet some of the CI failures look related (h3 quiche).  I'm re-running and will start reviewing as well

gregw commented on jetty-core/jetty-quic/jetty-quic-api/src/main/java/org/eclipse/jetty/quic/api/frames/Frame.java:
```diff
@@ -0,0 +1,59 @@
+    @Override
+    public String toString()
+    {
+        return "%s@%x".formatted(getClass().getSimpleName(), hashCode());
```
The new "fashion" in `toString` implementations is to use the short name for the class, this is probably a good thing to do for such a deep package.   I'm wondering if we should add a method to `TypeUtil` like:

```java
    public static String toString(Object subject, String format, Object... args)
    {
        String combinedFormat = "%s@%x{" + format + "}";
        Object[] allArgs = new Object[args.length + 2];
        allArgs[0] = TypeUtil.toShortName(subject.getClass());
        allArgs[1] = subject.hashCode();
        System.arraycopy(args, 0, allArgs, 2, args.length);
        return combinedFormat.formatted(allArgs);
    }
```

I know this is kind of orthogonal to this PR, but you are introducing a lot of new classes that will be toStringed.

sbordet replied:
Now using `TypeUtil`, but the larger change is for a different PR, if at all.

gregw commented on jetty-core/jetty-quic/jetty-quic-api/src/main/java/org/eclipse/jetty/quic/api/frames/Frame.java:
```diff
@@ -0,0 +1,59 @@
+
+package org.eclipse.jetty.quic.api.frames;
+
+public class Frame
```
I tiny amount of javadoc would be good to describe the style of frame class this is.  Is it immutable? Can payloads be consumed etc.

gregw commented on jetty-core/jetty-quic/jetty-quic-api/src/main/java/org/eclipse/jetty/quic/api/frames/StreamFrame.java:
```diff
@@ -0,0 +1,145 @@
+     * so that the implementation can compute the {@code offset} on behalf of the application.</p>
+     *
+     * @param streamId the stream id generated using {@link Session#newStreamId(boolean)}
+     * @param data the data bytes to send
```
send only?  is this ony client side?

sbordet replied:
Yes, send only, and yes, client-side only. This frame is to be used to "open" a stream and obtain the `Stream` object.

gregw commented on jetty-core/jetty-quic/jetty-quic-api/src/main/java/org/eclipse/jetty/quic/api/Stream.java:
```diff
@@ -0,0 +1,428 @@
+    /**
+     * <p>A {@link Retainable} wrapper for stream data bytes.</p>
+     */
+    interface Data extends Retainable
```
Why isn't this a RBB in the same way that `Content.Chunk` is now?  It is-a `Retainable` and has-a `ByteBuffer`, so it is-a `RetainableByteBuffer` and can benefit from all the extra methods and utilities that we have for RBBs

sbordet replied:
Removed, replaced by `Chunk`.

gregw commented on jetty-core/jetty-quic/jetty-quic-api/src/main/java/org/eclipse/jetty/quic/api/frames/StreamFrame.java:
```diff
@@ -0,0 +1,145 @@
+    }
+
+    private final long offset;
+    private final ByteBuffer data;
```
I'm looking for how retainability is handled for buffers used by the Stream?  I can see that the `Data` objects produced by the stream are retainable, but I can't see how that might be reflected here.   So I would expect at least a javadoc discussing how retainability is handled for these buffers... if not actual explicit retention here.

sbordet replied:
The frame is just a record, it only carries the data that has been parsed, or whatever the application creates.
Retainability is not relevant here, it is only relevant for `Stream.Data`.

gregw commented on jetty-core/jetty-quic/jetty-quic-api/src/main/java/org/eclipse/jetty/quic/api/frames/StreamFrame.java:
```diff
@@ -0,0 +1,145 @@
+
+    private final long offset;
+    private final ByteBuffer data;
+    private final int length;
```
obviously the length is the original length of the bytebuffer.  But need javadoc to discuss how/when the byte buffer is consumed and what it means if the remaining does not match this length

sbordet replied:
Consumption of the buffer is handled in `Stream.Data`.

gregw commented on jetty-core/jetty-quic/jetty-quic-api/src/main/java/org/eclipse/jetty/quic/api/frames/TransportParameters.java:
```diff
@@ -0,0 +1,182 @@
+        return "%s@%x[%s]".formatted(getClass().getSimpleName(), hashCode(), parameters);
+    }
+
+    public abstract static class Id<T>
```
javadoc

gregw commented on jetty-core/jetty-quic/jetty-quic-api/src/main/java/org/eclipse/jetty/quic/api/frames/TransportParameters.java:
```diff
@@ -0,0 +1,182 @@
+        @Override
+        public String toString()
+        {
+            return "%d".formatted(id);
```
this should be more than just the ID as it is possible to have two instances of the same ID

sbordet replied:
Made the constructor private, so applications cannot create instances.
If necessary for QUIC on Streams, we can have factory methods, but now are not necessary.

gregw commented on jetty-core/jetty-quic/jetty-quic-api/src/main/java/org/eclipse/jetty/quic/api/Session.java:
```diff
@@ -0,0 +1,202 @@
+    /**
+     * @return the QUIC connection id
+     */
+    String getId();
```
why not a `TransportParameter.Id`?  If not appropriate, why is that class public in the API?

sbordet replied:
`Session.id` is the unique connection id for that session, and is not related to `TransportParameters`.

`TransportParameters.Id` is public because applications should be able to configure transport parameters and possibly create custom ones.

gregw commented on jetty-core/jetty-quic/jetty-quic-api/src/main/java/org/eclipse/jetty/quic/api/Stream.java:
```diff
@@ -0,0 +1,428 @@
+     * or {@code null} if no data bytes are available
+     * @see #demand()
+     */
+    Data read();
```
Why is there no way to return errors from this method? I think we benefited a lot from the `Content.Source#read` contract, which can serialize failures with data arrival.  What happens if a call to `read` is racing with a call to `Listener#onFailure`?

I also see there is a `StreamDataFailure` record, that does have a failure in it, but nothing in the contract of read says to expect that might be returned?

sbordet replied:
Converted to return `Chunk`.

gregw commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/Retainable.java:
```diff
@@ -244,7 +244,7 @@ public int getRetained()
         @Override
         public String toString()
```
another candidate for a toString short name

gregw commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/RetainableByteBuffer.java:
```diff
@@ -935,7 +935,7 @@ protected void addStringInfo(StringBuilder builder)
             builder.append(isDirect());
             addExtraStringInfo(builder);
-            builder.append(",r=");
+            builder.append(",rc=");
```
short name above?

gregw commented on jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/ServletMultiPartFormData.java:
```diff
@@ -83,7 +83,7 @@ static Parts getParts(ServletRequest servletRequest)
     {
         CompletableFuture<Parts> futureParts = from(servletRequest, promise.getInvocationType(), contentType);
-        futureParts.whenComplete(promise);
+        futureParts.whenComplete(Promise.Invocable.toBiConsumer(promise));
```
We need to open an issue to get rid of the implementations using the deprecated CompletableFuture APIs

sbordet committed:
Merged branch 'jetty-12.1.x' into 'jetty-12.1.x-quic-streams'.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Merged branch 'jetty-12.1.x' into 'jetty-12.1.x-quic-streams'.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet commented:
@gregw yes, `quic.Stream.Data` can be a `Chunk`, and as such return a failure. I'll work on that.

However, we need a mechanism that delivers asynchronous failures also in case there is no demand, for example when the remote peer sends a RESET_STREAM, but the local application is not demanding for data (we cannot call a demand callback if there is no demand).

The same mechanism is used to deliver a failure that happens _before_ calling the application, for example a QPACK failure, or a protocol failure when receiving a HEADERS frame (e.g. a duplicate or non-existing stream id).

So the two mechanisms must co-exist, at least at this low-level API.

At the `Handler` level, we deliver this kind of asynchronous failures either via `Request.read()`, or via a `Request` failure listener (sometimes both), so we do have two mechanisms all the way up.

sbordet committed:
Updates from review.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Merged branch 'jetty-12.1.x' into 'jetty-12.1.x-quic-streams'.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Merged branch 'jetty-12.1.x' into 'jetty-12.1.x-quic-streams'.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Updates from review.

Removed quic.Stream.Data and http3.Stream.Data classes, replaced by Content.Chunk.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Fixed test relying on toString().

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

gregw commented on jetty-core/jetty-quic/jetty-quic-api/src/main/java/org/eclipse/jetty/quic/api/Stream.java:
```diff
@@ -0,0 +1,376 @@
+ *
+ * @see Stream.Listener
+ */
+public interface Stream
```
So why isn't this a `Content.Source`?   See further comments below...

gregw commented on jetty-core/jetty-quic/jetty-quic-api/src/main/java/org/eclipse/jetty/quic/api/Stream.java:
```diff
@@ -0,0 +1,376 @@
+     * or a failure, or {@code null} if no data bytes are available
+     * @see #demand()
+     */
+    Content.Chunk read();
```
this is the same contract as `Content.Source#read()`

gregw commented on jetty-core/jetty-quic/jetty-quic-api/src/main/java/org/eclipse/jetty/quic/api/Stream.java:
```diff
@@ -0,0 +1,376 @@
+         * @param failure the idle timeout failure
+         * @return whether to notify the other peer that this stream is closing
+         */
+        default boolean onIdleTimeout(Stream stream, TimeoutException failure)
```
How is this different to reading an error chunk for idle timeout?   Can idle be non-terminal?

sbordet replied:
When no data demand, this method is called. For example, when all the content has been already read.
It can be non-terminal -- the boolean indicates whether the stream should be closed or kept open.

gregw commented on jetty-core/jetty-quic/jetty-quic-api/src/main/java/org/eclipse/jetty/quic/api/Stream.java:
```diff
@@ -0,0 +1,376 @@
+         * @param stream the stream
+         * @param failure the stream failure
+         */
+        default void onFailure(Stream stream, Throwable failure)
```
You need to document how this is different to reading a failure and if a failure will be reported to both.    All the thinking we did about the same listeners in `org.eclipse.jetty.server.internal.HttpChannelState.onFailure(java.lang.Throwable, boolean)` needs to be repeated here.

gregw commented on jetty-core/jetty-quic/jetty-quic-api/src/main/java/org/eclipse/jetty/quic/api/Stream.java:
```diff
@@ -0,0 +1,376 @@
+         *
+         * @param stream the stream
+         */
+        default void onClose(Stream stream)
```
is this called before or after an EOF chunk is read?

sbordet replied:
Updated javadocs, clarifying when this method is called.

gregw reviewed (changes requested):
I'm concerned that this PR might repeat some mistakes we made in HttpChannel with regards to asynchronous failures, timeouts and reading chunks.      Would it not be better to make `Stream` a `Content.Source`, so we could apply the same logic?  I know there are also additional semantics, but surely they can be laid on top?

gregw commented on jetty-core/jetty-quic/jetty-quic-api/src/main/java/org/eclipse/jetty/quic/api/Stream.java:
```diff
@@ -0,0 +1,376 @@
+     * @see #read()
+     * @see Listener#onDataAvailable(Stream)
+     */
+    void demand();
```
There is nothing in this abstract contract that allows a Listener to be associated with a Stream?

gregw replied:
Hmmm now I notice that `HttpStream` is-not-a `Content.Source`.... hmmmm

gregw commented on jetty-core/jetty-http3/jetty-http3-client/src/main/java/org/eclipse/jetty/http3/client/HTTP3SessionClient.java:
```diff
@@ -91,49 +96,53 @@ public void onSettings(SettingsFrame frame)
         ProtocolSession session = getProtocolSession();
-        QuicStreamEndPoint endPoint = session.getOrCreateStreamEndPoint(streamId, session::openProtocolEndPoint);
+        StreamEndPoint endPoint = session.createStreamEndPoint(quicStream, session::openStreamEndPoint);
+        ((AbstractStream)quicStream).setListener(new ProtocolStreamListener(() -> endPoint));
```
setting the listener a few lines after the construction looks sub-optimal.  OK, the stream needs to know the endPoint and the endPoint needs to know the stream, but that feels like there should be a setter for endPoint rather than for the listener.

sbordet committed:
Updated TransportParameters for future usage.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Updated Stream.Listener.onIdleTimeout() to use a Promise, rather than returning a boolean.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Fixed case when the trailer HEADERS frame is QPACK-blocked.

HTTP upper layers (core, servlet) may insist reading even if null was already returned to them because the trailer is QPACK-blocked.

Trying to read more was delegated to the QUIC layer, but for QUIC the stream is complete, since it has delivered the trailer HEADERS frame; the actual headers arrive on the HTTP/3 encoder stream, but it's a different stream.

Now trying to read more will return null until the stream is QPACK-unblocked.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

gregw commented:
@sbordet you have conflicts to resolve and `testResetAfterTCPCongestedWriteThenResolveCongestionToUnblockWrite – org.eclipse.jetty.http2.tests.StreamResetTest` looks at the very least 50% flakey.
If you can fix those, then lets get this merged!

lorban committed:
Merge remote-tracking branch 'origin/jetty-12.1.x' into fix/12.1.x/quiche-rewrite

lorban commented:
@sbordet @gregw I've taken care of the merge conflict since it was my quiche upgrade that caused it.

sbordet committed:
Merged branch 'jetty-12.1.x' into 'jetty-12.1.x-quiche-rewrite'.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Update HTTP3Client to not modify the ClientQuicConfiguration passed to the constructor.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet commented:
Merged in https://github.com/jetty/jetty.project/commit/541c3ee9d88ef09614744ce1641262041d7620dd.
